src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,432 @@
+end
+
+# image(f, a) done by parent call overloading
+
+function preimage(f::OscarSingularCoefficientRingMapGeneric, a::Singular.n_unknown)
+  parent(a) !== codomain(f) && error("Element not in codomain")
+  b = Singular.libSingular.julia(Singular.libSingular.cast_number_to_void(a.ptr))
+  return b::elem_type(domain(f))
+end
+
+# ZZ
+iso_oscar_singular_coeff_ring(R::ZZRing) = OscarSingularCoefficientRingMapGeneric(R, Singular.Integers())
+
+# QQ
+iso_oscar_singular_coeff_ring(R::QQField) = OscarSingularCoefficientRingMapGeneric(R, Singular.Rationals())
+
+# prime field, small characteristic
+iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapGeneric(R, Singular.Fp(Int(characteristic(R))))
+
+# ZZ/nZZ, n small and big
+iso_oscar_singular_coeff_ring(R::Union{zzModRing, ZZModRing}) = OscarSingularCoefficientRingMapGeneric(R, Singular.residue_ring(Singular.Integers(), BigInt(modulus(R)))[1]
+)
+
+# QQ(a)
+function iso_oscar_singular_coeff_ring(R::AbsSimpleNumField)
+  minpoly = defining_polynomial(R)
+  Qa = parent(minpoly)
+  a = gen(Qa)
+  SQa, (Sa,) = Singular.FunctionField(Singular.QQ, _variables_for_singular(symbols(Qa)))
+  Sminpoly = SQa(coeff(minpoly, 0))
+  for i in 1:degree(minpoly)
+    Sminpoly += SQa(coeff(minpoly, i))*Sa^i
+  end
+  SK, _ = Singular.AlgebraicExtensionField(SQa, Sminpoly)
+  return OscarSingularCoefficientRingMapGeneric(R, SK)
+end
+
+# Conversion from GF(p, n), small p, to Singular.N_AlgExtField,
+# the code for the conversion is present in Singular.jl/src/number/n_algExt.jl
+# via parent object call overloading, so we wrap it here anyway.
+
+# GF(p, n), small p
+function iso_oscar_singular_coeff_ring(F::fqPolyRepField)
+  # TODO: the Fp(Int(char)) can throw
+  minpoly = modulus(F)
+  Fa = parent(minpoly)
+  SFa, (Sa,) = Singular.FunctionField(Singular.Fp(Int(characteristic(F))),
+                                                    _variables_for_singular(symbols(Fa)))
+  Sminpoly = SFa(coeff(minpoly, 0))
+  for i in 1:degree(minpoly)
+    Sminpoly += SFa(coeff(minpoly, i))*Sa^i
+  end
+  SF, _ = Singular.AlgebraicExtensionField(SFa, Sminpoly)
+  return OscarSingularCoefficientRingMapGeneric(F, SF)
+end
+
+# Finite field (FqField)
+struct OscarSingularCoefficientRingMapFqField{D} <: OscarSingularCoefficientRingMap{FqField, D}
+  R::FqField
+  S::D
+  iso::MapFromFunc{FqField, FqField}
+
+  OscarSingularCoefficientRingMapFqField(R, S) = new{typeof(S)}(R, S)
+  OscarSingularCoefficientRingMapFqField(R, S, iso) = new{typeof(S)}(R, S, iso)
+end
+
+function _absolute_field(F::FqField)
+  if is_absolute(F)
+    error("don't use me, you are already absolute")
+  end
+  Kx = parent(defining_polynomial(F))
+  Fabs, QQtoFabs = Nemo._residue_field(defining_polynomial(F); absolute = true, check = false)
+  Fabsx, = polynomial_ring(Fabs, :x; cached = false)
+  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), b -> begin a = Fabs(); Nemo.set!(a, b); a end)
```

Comment:
For which types does `Nemo.set!(a, b)` work but not `Fabs(b)`?

otherwise we could do
```suggestion
  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), b -> Fabs(b))
```
Or even
```suggestion
  return Fabs, MapFromFunc(Fabs, F, a -> F(preimage(QQtoFabs, a)), Fabs)
```

Reply:
It is not a matter of types. This is reinterpreting elements between different flint finite fields and outside the scope of "promotion".